### PR TITLE
Add SpatialInertia::SolidCylinderAboutEnd()

### PR DESCRIPTION
**Pull request by mitiguy**
This change is [Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18889)

### CODE REVIEW

**Comment by mitiguy** [comment]
PR https://github.com/RobotLocomotion/drake/pull/18889 is work to helping resolve part of issue https://github.com/RobotLocomotion/drake/issues/18416.

**Review by mitiguy — COMMENTED** [review]
Feature review +@SeanCurtis-TRI 

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18889#-:-NPL-TQI1jcYiBUmCIlX:b-j0fgx5)* status: LGTM missing from assignee SeanCurtis-TRI(platform), needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @SeanCurtis-TRI)

**Review by SeanCurtis-TRI — COMMENTED** [review]
:LGTM: With a single editorial nit (spread over two places).

+@rpoyner-tri for platform review.

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18889#-:-NPNp4Gv6PbISIT4g1fG:boclug2)* status: 5 unresolved discussions, LGTM missing from assignee rpoyner-tri(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @mitiguy and @rpoyner-tri)

___
*[`multibody/tree/spatial_inertia.h` line 179 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/18889#-NPNlOZFBwSMAQaAHdWn:-NPNlOZFBwSMAQaAHdWo:b-cgcca4) ([raw file](https://github.com/RobotLocomotion/drake/blob/683da664be1cd6b2e4da56f0c885722e40175a87/multibody/tree/spatial_inertia.h#L179)):*
> ```objc
> 
>   /// Creates a spatial inertia for a uniform density solid cylinder B about its
>   /// end-point Bp (Bp is at the center of a circular end of the cylinder).
> ```

nit: This is ambiguous. The use of the word "the" implies a uniquely defined point, but I can think of two points on the cylinder that satisfy this description.

You *do* give a more formal definition below, so perhaps a slight rephrasing (replacing definite articles with indefinite articles and explicitly letting them know the details are found later in the docs) might happily resolve it:



_Suggestion:_
```objc
  /// Creates a spatial inertia for a uniform density solid cylinder B about a
  /// point Bp (located at the center of one of the cylinder's circular ends;
  /// see below for more details).
```
___
*[`multibody/tree/spatial_inertia.h` line 186 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/18889#-NPNmjri385xC7BQnZNz:-NPNmjri385xC7BQnZO-:b-d0m4no) ([raw file](https://github.com/RobotLocomotion/drake/blob/683da664be1cd6b2e4da56f0c885722e40175a87/multibody/tree/spatial_inertia.h#L186)):*
> ```objc
>   ///   cylinder, expressed in B.
>   /// @retval M_BBp_B B's spatial inertia about Bp, expressed in B.
>   /// @note The position from Bp to Bcm is p_BpBcm_B = 0.5*l*unit_vector.
> ```

BTW It's not clear why you're defining p_BpBcm here in the documentation when the code uses p_BcmBp (likewise the tests). Is the assumption that p_BpBcm will be of extra use/significance to the caller (even though the APIs being exercised requires the opposite)?

If there is no particular benefit to the caller of this API, I'd suggest reversing it so the documentation and code more obviously match. It'll make it easier for people who have to read the code.
___
*[`multibody/tree/spatial_inertia.h` line 187 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/18889#-NPNmHvq9mdSNxwKSmI6:-NPNmHvq9mdSNxwKSmI7:b-ajqije) ([raw file](https://github.com/RobotLocomotion/drake/blob/683da664be1cd6b2e4da56f0c885722e40175a87/multibody/tree/spatial_inertia.h#L187)):*
> ```objc
>   /// @retval M_BBp_B B's spatial inertia about Bp, expressed in B.
>   /// @note The position from Bp to Bcm is p_BpBcm_B = 0.5*l*unit_vector.
>   /// @note B's rotational inertia about Bp is axially symmetric, meaning B has
> ```

BTW I'm not quite sure what the caller is supposed to do with this information. Can you shed some light on that?
___
*[`multibody/tree/spatial_inertia.cc` line 117 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/18889#-NPNnM798xlaa5b0kcat:-NPNnM798xlaa5b0kcau:b93l2lx) ([raw file](https://github.com/RobotLocomotion/drake/blob/683da664be1cd6b2e4da56f0c885722e40175a87/multibody/tree/spatial_inertia.cc#L117)):*
> ```cpp
>   const Vector3<T> p_BcmBp_B = -0.5 * l * unit_vector;
>   const SpatialInertia<T> M_BBp_B = M_BBcm_B.Shift(p_BcmBp_B);
>   return M_BBp_B;
> ```

BTW In `SolidTetrahedronAboutPointWithDensity()` there was an extended discussion about how many copies are being done. There, we've gone out of our way to *not* incur an extra copy. Here we've implicitly said, "meh, let's go ahead and copy."

In general, the extra copy is not a terrible thing -- I presume that this is an initialization-type function and the cost of the copy is negligible compared to all of the operations that will be done on the resulting data. So, I don't have a complaint about a copy happening here for any principle-based reason.

But, given the contortions on the other API, I wonder about the *consistency* of philosophy being applied in this family of functions. I'd suggest they should match (with good arguments that the tetrahedron method should shrug and do a copy as well because it makes for better self-documenting code).
___
*[`multibody/tree/test/spatial_inertia_test.cc` line 194 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/18889#-NPNoMLQ8SJcg070nrFW:-NPNoMLQ8SJcg070nrFX:b-hrfluh) ([raw file](https://github.com/RobotLocomotion/drake/blob/683da664be1cd6b2e4da56f0c885722e40175a87/multibody/tree/test/spatial_inertia_test.cc#L194)):*
> ```cpp
>       CompareMatrices(M_expected.CopyToFullMatrix6(), M.CopyToFullMatrix6()));
> 
>   // Test a solid cylinder B about Bp, where Bp is at the center of the circular
> ```

nit More of the definite article "the" to describe an object that is not unique.

**Review by rpoyner-tri — COMMENTED** [review]
:lgtm:

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18889#-:-NPNuqRl7hAybNJMNW5h:bnfp4nl)* status: 5 unresolved discussions, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @mitiguy)

**Review by mitiguy — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18889)* status: 2 unresolved discussions (waiting on @rpoyner-tri and @SeanCurtis-TRI)

___
*[`multibody/tree/spatial_inertia.h` line 179 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/18889#-NPNlOZFBwSMAQaAHdWn:-NPPF5D1BShEI8jS44u-:bkiv9bo) ([raw file](https://github.com/RobotLocomotion/drake/blob/683da664be1cd6b2e4da56f0c885722e40175a87/multibody/tree/spatial_inertia.h#L179)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

nit: This is ambiguous. The use of the word "the" implies a uniquely defined point, but I can think of two points on the cylinder that satisfy this description.

You *do* give a more formal definition below, so perhaps a slight rephrasing (replacing definite articles with indefinite articles and explicitly letting them know the details are found later in the docs) might happily resolve it:
</blockquote></details>

Done. Changed "its" to "an" to avoid the definite/indefinite clash and integrated your suggestion.
___
*[`multibody/tree/spatial_inertia.h` line 186 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/18889#-NPNmjri385xC7BQnZNz:-NPPFOYBC8DfQdiADtEk:b291r6u) ([raw file](https://github.com/RobotLocomotion/drake/blob/683da664be1cd6b2e4da56f0c885722e40175a87/multibody/tree/spatial_inertia.h#L186)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW It's not clear why you're defining p_BpBcm here in the documentation when the code uses p_BcmBp (likewise the tests). Is the assumption that p_BpBcm will be of extra use/significance to the caller (even though the APIs being exercised requires the opposite)?

If there is no particular benefit to the caller of this API, I'd suggest reversing it so the documentation and code more obviously match. It'll make it easier for people who have to read the code.
</blockquote></details>

Done.  I removed the monogram notation as the words and mathematical equations are explicit and suffice.  
Note: The expressed-in frame is unambiguous due to the documentation for unit_vector.
___
*[`multibody/tree/spatial_inertia.h` line 187 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/18889#-NPNmHvq9mdSNxwKSmI6:-NPPFnhsARd8iWoTqAtJ:b-p47s9g) ([raw file](https://github.com/RobotLocomotion/drake/blob/683da664be1cd6b2e4da56f0c885722e40175a87/multibody/tree/spatial_inertia.h#L187)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW I'm not quite sure what the caller is supposed to do with this information. Can you shed some light on that?
</blockquote></details>

This is purely an FYI that may (or may not) be helpful to the user.  
In other words, the user should not think it a mistake if two of the three principal moments of inertia are equal.
___
*[`multibody/tree/spatial_inertia.cc` line 117 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/18889#-NPNnM798xlaa5b0kcat:-NPPGtJIATfxkQ0az4L9:b4o7elc) ([raw file](https://github.com/RobotLocomotion/drake/blob/683da664be1cd6b2e4da56f0c885722e40175a87/multibody/tree/spatial_inertia.cc#L117)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW In `SolidTetrahedronAboutPointWithDensity()` there was an extended discussion about how many copies are being done. There, we've gone out of our way to *not* incur an extra copy. Here we've implicitly said, "meh, let's go ahead and copy."

In general, the extra copy is not a terrible thing -- I presume that this is an initialization-type function and the cost of the copy is negligible compared to all of the operations that will be done on the resulting data. So, I don't have a complaint about a copy happening here for any principle-based reason.

But, given the contortions on the other API, I wonder about the *consistency* of philosophy being applied in this family of functions. I'd suggest they should match (with good arguments that the tetrahedron method should shrug and do a copy as well because it makes for better self-documenting code).
</blockquote></details>

Agreed and done.
___
*[`multibody/tree/test/spatial_inertia_test.cc` line 194 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/18889#-NPNoMLQ8SJcg070nrFW:-NPPHZPL9mrzxlsInhOE:b-896fix) ([raw file](https://github.com/RobotLocomotion/drake/blob/683da664be1cd6b2e4da56f0c885722e40175a87/multibody/tree/test/spatial_inertia_test.cc#L194)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

nit More of the definite article "the" to describe an object that is not unique.
</blockquote></details>

Done.

**Review by SeanCurtis-TRI — COMMENTED** [review]
> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18889)* status: :shipit: complete! all discussions resolved, LGTM from assignees rpoyner-tri(platform),SeanCurtis-TRI(platform) (waiting on @mitiguy)

___
*[`multibody/tree/spatial_inertia.h` line 187 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/18889#-NPNmHvq9mdSNxwKSmI6:-NPPmzpiDy1PDA4SR2gc:bh4r4cs) ([raw file](https://github.com/RobotLocomotion/drake/blob/683da664be1cd6b2e4da56f0c885722e40175a87/multibody/tree/spatial_inertia.h#L187)):*
<details><summary><i>Previously, mitiguy (Mitiguy) wrote…</i></summary><blockquote>

This is purely an FYI that may (or may not) be helpful to the user.  
In other words, the user should not think it a mistake if two of the three principal moments of inertia are equal.
</blockquote></details>

I'm not sure that the phrase "two of the three principal moments of inertia are equal" is easily mapped to the text that is written here by any except the most rarefied of experts.

**Review by mitiguy — COMMENTED** [review]
+(status: squashing now) 

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18889#-:-NPPoLET-biRRJV0m-_r:bm9o57a)* status: :shipit: complete! all discussions resolved, LGTM from assignees rpoyner-tri(platform),SeanCurtis-TRI(platform) (waiting on @mitiguy)
